internal/postgres/data.go

```diff
@@ -0,0 +1,266 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package postgres
+
+import (
+	"encoding/hex"
+	"fmt"
+	"harbourbridge/spanner/ddl"
+	"math/bits"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/civil"
+)
+
+// ConvertData maps the PostgreSQL data in vals into Spanner data,
+// based on the specified Spanner and PostgreSQL schemas. Note that since entries
+// in vals may be empty, we also return the list of columns (empty cols are dropped).
+func ConvertData(conv *Conv, spTable, pgTable string, cols []string, vals []string) ([]string, []interface{}, error) {
+	spannerSchema, ok1 := conv.spSchema[spTable]
+	pgSchema, ok2 := conv.pgSchema[pgTable]
+	if !ok1 || !ok2 {
+		return []string{}, []interface{}{}, fmt.Errorf("Can't find table %s in schema", spTable)
+	}
+	var c []string
+	var v []interface{}
+	if len(cols) != len(vals) {
+		return []string{}, []interface{}{}, fmt.Errorf("Bad parameters: cols and vals have different lengths: len(cols)=%d, len(vals)=%d", len(cols), len(vals))
+	}
+	for i, col := range cols {
+		if vals[i] == "\\N" { // PostgreSQL representation of empty column in COPY-FROM blocks.
+			continue
+		}
+		cd, ok := spannerSchema.Cds[col]
+		if !ok {
+			return []string{}, []interface{}{}, fmt.Errorf("Can't find col %s in schema", col)
+		}
+		pgCol, ok := pgSchema.cols[col]
+		if !ok {
+			return []string{}, []interface{}{}, fmt.Errorf("Can't find col %s in pgSchema", col)
+		}
+		var x interface{}
+		var err error
+		if cd.IsArray {
+			x, err = convArray(cd.T, pgCol, conv.location, vals[i])
+		} else {
+			x, err = convScalar(cd.T, pgCol.id, conv.location, vals[i])
+		}
+		if err != nil {
+			return []string{}, []interface{}{}, err
+		}
+		v = append(v, x)
+		c = append(c, col)
+	}
+	if aux, ok := conv.syntheticPKeys[spTable]; ok {
+		c = append(c, aux.col)
+		v = append(v, int64(bits.Reverse64(uint64(aux.sequence))))
+		aux.sequence++
+		conv.syntheticPKeys[spTable] = aux
+	}
```

Comment:
what is syntheticPKeys for?

Reply:
Spanner requires primary keys for all tables. PostgreSQL recommends the use of
primary keys for all tables, but does not enforce this. When converting a PostgreSQL table
without a primary key, we need to create a new primary key -- these are called synthetic keys.
They are introduced in statement.go (see https://github.com/cloudspannerecosystem/harbourbridge/pull/11).
We need to deal with them in data because we need to generate values for these new columns (since
they don't appear in the PostgreSQL data).

Reply:
Thx for the explanation.